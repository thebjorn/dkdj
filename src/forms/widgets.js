/*
 *  dk.forms namespace
 *
 *  The purpose of these widgets is to auto-create widgets for inline editing of grids.
 */

import dk from "../dk-obj";
import is from "../is";
import {Duration} from "../data/datacore/dk-datatypes";
import {env} from "../lifecycle/dkboot/lifecycle-parse-script-tag";
import {InputWidget} from "./input-widget";
import {dedent} from "../text/template-functions";
import {dkwarning} from "../lifecycle/coldboot/dkwarning";
import {dkconsole} from "../lifecycle/dkboot/dk-console";
import {zip_object} from "../collections";

/*
    Global attributes:
    
    accesskey, class
    contenteditable
    contextmenu-x
    data-*
    dir
    draggable
    dropzone-x
    hidden
    id
    inputmode-x
    is
    itemid, itemprop, itemref, itemscope, itemtype
    lang
    slot
    spellcheck
    style
    tabindex
    title
    translate-x
    
 */


export class TextInputWidget extends InputWidget {
    constructor(...args) {
        super({
            template: {root: 'input'}
        }, ...args);
    }
    construct() {
        super.construct();
        if (this.node.type !== 'text') this.node.type = 'text';
    }
    handlers() {
        super.handlers();
        this.retrigger('keyup');
    }
}


export class DurationWidget extends InputWidget {
    constructor(...args) {
        super({
            template: {root: 'input'},
        }, ...args);
    }
    construct() {
        super.construct();
        if (this.node.type !== 'text') this.node.type = 'text';
    }
    parse(str) {
        if (str instanceof Duration) return str;
        return new Duration(str);
    }
    stringify(val) {
        if (val instanceof Duration) return val.toString();
        return (new Duration(val)).toString();
    }

}


export class RadioInputWidget extends InputWidget {
    constructor(...args) {
        const props = Object.assign({}, ...args);
        const _checked = props.checked || false;
        delete props.checked;
        
        super({
            template: {root: 'input'}
        }, ...args);
        this._checked = _checked;
    }
    
    construct() {
        super.construct();
        if (this.node.type !== 'radio') this.node.type = 'radio';
        this.node.checked = this._checked;
    }
    
    get checked() {
        return this._checked;
    }

    set checked(value) {
        this._checked = value;
        this.node.checked = value;  // XXX: should this be in set_node_value?
        if (env.DEBUG) this.node.setAttribute('checked', 'checked');
    }

    handlers() {
        const self = this;
        super.handlers();
        this.widget().on('change', function () {
            self._checked = self.node.checked;  // don't use setter here.
        });
    }
}

/* html attributes

        autocomplete, autofocus
        disabled
        form
        multiple
        name
        required
        size
        
 */

/**
 * create an options dict `{name -> value} from either
 *
 *  (a) an options dict,
 *  (b) a flat array, or
 *  (c) a [[key, val], ..] array.
 *
 * @param options - key value dict
 * @param validate - optional validation function
 */
function create_options(options, validate) {
    if (options == null) return {};
    let result = {};
    // let order = [];
    
    if (Array.isArray(options)) {  // (b) or (c)
        // https://www.stefanjudis.com/today-i-learned/property-order-is-predictable-in-javascript-objects-since-es2015/
        // property order in objects is by insertion order..
        // ..except for numbers -- which are sorted and inserted
        // in front (thus messing up an implied order that is 
        // not numerically ascending)
        if (options.length > 0) {  // options === []
            let [first, ...rest] = options;
            if (Array.isArray(first)) {  // (c) options === [[k,v], [k,v],...]
                options.forEach(([k, v]) => result[k] = v);
                // order = options.map(([k, v]) => k);
            } else {   // optinos === [v1, v2, v3, ...]
                // order = options;
                options.forEach(v => result[v] = v);
                if (!is.isEqual(Object.keys(result), options.map(v => `${v}`))) {
                    throw `The default ordering of the options 
                     ${JSON.stringify(Object.keys(result))} does not match ${JSON.stringify(options)}. You
                     should specify the options as a property object (instead
                     of a list) to achieve your desired order.`;
                }
                // options.forEach(v => result[`${(typeof v === 'number' ? '0' : '')}${v}`] = v);
            }
        }
    } else {  // (a)
        result = options;
        // order = Object.keys(options);
    }

    if (validate) {
        Object.keys(result).forEach(k => {
            if (result[k] !== validate[k]) {
                throw dedent`
                    Illegal value (${k}) - not in options
                    VALUES: ${JSON.stringify(result)}
                    VALIDATE: ${JSON.stringify(validate)} 
                `;
            }
        });
    }
    return result;
}


export class SelectWidget extends InputWidget {
    constructor(...args) {
        // console.log("ARGS:", args);
        const props = Object.assign({}, ...args);
        const options = create_options(props.options);
        delete props.options;
        
        const value = create_options(props.value, options);
        delete props.value;
        
        super({
            size: 1,
            multiple: false,
            data: {value: value},
            template: {root: 'select'},
        }, props);

        this._selected = {};
        this.options = options;
        if (!this.multiple && Object.keys(value).length > 1) throw "cannot have multiple values when multiple=false";
        Object.keys(value).forEach(k => this._selected[k] = true);
    }
    get options() { return this._options; }
    set options(options) {
        this._options = create_options(options);
        Object.keys(this._options).forEach(k => this._selected[k] = false);
    }

    get value() {
        // make a real array from the keys
        const res = [];
        Object.keys(this.data.value).forEach(k => res.push(k));
        return res;
    }
    set value(v) {
        if (!Array.isArray(v)) v = [v];
        if (!this.multiple) {
            if (v.length > 1) throw "Cannot set multiple values when multiple = false!";
        }
        Object.keys(this._selected).forEach(k => this._selected[k] = false);
        v.forEach(k => this._selected[k] = true);
        return this._set_value_from_selected();
    }
    _get_value_from_selected() {
        // const val = [];
        // Object.entries(this._selected).forEach(([k, v]) => {
        //     if (v) val.push(k);
        // });
        // return val;
        const val = {};
        Object.entries(this._selected).forEach(([k, v]) => {
            if (v) val[k] = this.options[k];
        });
        return val;
    }
    _set_value_from_selected() {
        this.data.value = this._get_value_from_selected();
        return this.data.value;
    }
    
    set dom_value(v) {
        ++this._updating;
        this.widget().val(v);
        --this._updating;
    }
    
    rebuild_options() {
        const widget = this.widget();
        widget.empty();
        Object.entries(this.options).forEach(([attr, value]) => {
            const option = dk.$('<option/>').val(attr).text(value);
            // noinspection EqualityComparisonWithCoercionJS
            if (attr == this.value) option.attr('selected', 'selected');  // we want 1 == "1" here!
            widget.append(option);
            widget.append('\n    ');
        });
    }
    
    construct() {
        super.construct();
        if (this.options) this.rebuild_options();
    }

    draw(data) {
        if (data != null) {
            if (data.options && !is.isEqual(data.options, this.options)) {
                this.options = data.options;
                this.rebuild_options();
            } else {
                this.dom_value = data.value;
            }
        } else {
            this.dom_value = this.value;
        }
    }
    formatted_value() {
        return this.options[this.value];
    }
    get_field_value() {
        return {v: this.value, f: this.formatted_value()};
    }
}


export class RadioSelectWidget extends SelectWidget {
    constructor(...args) {
        super({
            template: {root: 'div'},
        }, ...args);
    }
    set dom_value(v) {
        if (v == null) return;
        ++this._updating;
        const value = v.value || v.v || v;
        Object.keys(value).forEach(val => {
            this.widget(`:radio[value="${val}"]`).prop("checked", true);
        });
        --this._updating;
    }
    rebuild_options() {
        const widget = this.widget();
        widget.empty();
        Object.entries(this.options).forEach(([attr, value]) => {
            const radio = this.layout.make('input', {
                type: 'radio',
                name: this.name,
                id: dk.counter(this.name + '-')
            });
            const label = this.layout.make('label', {
                id: radio.prop('id') + '-label',
                'for': radio.prop('id')
            }).text(value);
            radio.val(attr);
            // noinspection EqualityComparisonWithCoercionJS
            if (this.value == attr) radio.prop('checked', true);  // we want 1 == "1" here
            label.prepend(radio);
            this.widget().append(label);
            this.widget().append('\n        ');
        });
    }
    handlers() {
        const self = this;
        this.widget().on('change', ':radio', function (e) {
            self.widget_changed({
                type: 'change',
                event: e,
                item: this
            });
            self.trigger('change', e, self);
        });
        this.retrigger('validation-change');
    }
}


export class CheckboxSelectWidget extends RadioSelectWidget {
    constructor(...args) {
        super({
            multiple: true
        }, ...args);
    }
    
    set dom_value(v) {
        if (v == null) return;
        ++this._updating;
        const value = v.value || v.v || v;
        this.widget(`:checkbox`).prop("checked", false);
        // this.widget(`:checkbox`).removeAttr("checked");
        Object.keys(value).forEach(val => {
            this.widget(`:checkbox[value="${val}"]`).prop("checked", true);
            // this.widget(`:checkbox[value="${val}"]`).attr("checked", "checked");
        });
        --this._updating;
    }
    rebuild_options() {
        console.info("REBUILD:OPTIONS:::::::::::::::::::::::")
        const widget = this.widget();
        const self = this;
        widget.empty();
        Object.entries(this.options).forEach(([attr, value]) => {
            console.log("SELF:THIS:", self === this);
            console.log("SELF:WIDGET:", self.widget());
            console.log("THIS:WIDGET:", this.widget());
            const chkbx = this.layout.make('input', {
                type: 'checkbox',
                name: this.name,
                id: dk.counter(this.name + '-')
            });
            const label = this.layout.make('label', {
                id: chkbx.prop('id') + '-label',
                'for': chkbx.prop('id')
            }).text(value);
            chkbx.val(attr);
            // noinspection EqualityComparisonWithCoercionJS
            console.info("THIS:VALUE:", this.value, "ATTR:", attr, "INCLUDES:", this.value.includes(attr));
            // if (this.value == attr) {  // / we want 1 == "1" here
            if (this.value.includes(attr)) {  // / we want 1 == "1" here
                chkbx.prop('checked', true);
                // chkbx.attr('checked', 'checked');
            }
            label.prepend(chkbx);
            widget.append(label);
            widget.append('\n');
        });
        console.log("REBUILD:OPTIONS:WIDGET:CHECKED:", this.widget().find(':checked').length);
    }
   
    widget_changed(event) {
        console.debug("WIDGET:CHANGED:", this.value, event.item.value, event.item.checked, this._updating);
        if (this._updating++ === 0 && event.type === 'change') {
            console.debug("SELECTED:BEFORE:", this._selected);
            this._selected[event.item.value] = event.item.checked;
            console.debug("SELECTED:AFTER:", this._selected);
            this._set_value_from_selected();
            console.debug("WIDGET:CHANGED:VALUE:", this.value);
        }
        --this._updating;
    }
    
    handlers() {
        const self = this;
        this.widget().on('change', ':checkbox', function (e) {
            self.widget_changed({
                type: 'change',
                event: e,
                item: this
            });
            self.trigger('change', e, self);
        });
        this.retrigger('validation-change');
    }
}


// XXX: is this a form widget?
export class TriboolWidget extends InputWidget {
    constructor(...args) {
        super({
            values: { 0: 0, 1: 1, 2: 2 },
            value: 0,
            structure: {  // basic layout of widget
                tribool_label_box: {},
                true_widget_box: {
                    css: { float: 'left' }
                },
                false_widget_box: {
                    css: { float: 'left' }
                },
            }
        }, ...args);
    }
    construct() {
        super.construct();
        // fixme: add to structure
        this.label = this.widget('label');
        this.tribool_label_box.text(this.label.text);
        this.true_widget = dk.$('<button type="button" class="btn btn-default btn-circle"><dk-icon src="check checked"></dk-icon></button>');
        this.true_widget_box.append(this.true_widget);
        this.false_widget = dk.$('<button type="button" class="btn btn-default btn-circle"><dk-icon src="minus checked"></dk-icon></i></button>');
        this.false_widget_box.append(this.false_widget);
    }
    set dom_value(v) {
        
    }
    
    draw(value) {
        if (value === undefined) return;

        switch (value) {
            case 0:
                this.true_widget.removeClass('btn-success').addClass('btn-default');
                this.false_widget.removeClass('btn-danger').addClass('btn-default');
                break;
            case 1:
                //this.true_widget.toggleClass('btn-default btn-success');
                this.true_widget.removeClass('btn-default').addClass('btn-success');
                this.false_widget.removeClass('btn-danger').addClass('btn-default');
                break;
            case 2:
                //this.false_widget.toggleClass('btn-default btn-danger');
                this.true_widget.removeClass('btn-success').addClass('btn-default');
                this.false_widget.removeClass('btn-default').addClass('btn-danger');
                break;
        }
    }
    handlers() {
        const self = this;
        this.true_widget.on('click', function () {
            switch (self.value) {
                case 0: self.value = 1; break;
                case 1: self.value = 0; break;
                case 2: self.value = 1; break;
            }
            self.draw(self.value);
            dk.trigger(self.widget, 'change', self.widget);
            self.notify('change', self);
        });
        this.false_widget.on('click', function() {
            switch (self.value) {
                case 0: self.value = 2; break;
                case 1: self.value = 2; break;
                case 2: self.value = 0; break;
            }
            self.draw(self.value);
            dk.trigger(self.widget, 'change', self.widget);
            self.notify('change', self);
        });
    }
}
