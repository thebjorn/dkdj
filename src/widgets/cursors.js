

const _default32 = {
    up: 'url(//static.datakortet.no/dk/sort-za-32.cur), pointer',
    down: 'url(//static.datakortet.no/dk/sort-az-32.cur), pointer'
};

const _default24 = {
    up: 'url(//static.datakortet.no/dk/sort-za.cur), pointer',
    down: 'url(//static.datakortet.no/dk/sort-az.cur), pointer'
};

export function cursor(name) {
    return _default24[name] || name;
}
