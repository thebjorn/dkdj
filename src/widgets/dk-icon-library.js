import dk from "../dk-obj";
import {dkrequire} from "../lifecycle/browser/dk-require";
import dom from "../browser/dom";


// xxx: should this be exported..?
export class IconLibrary {
    constructor(props) {
        Object.assign(this, {
            url: '',
            prefix: '',
            classes: '',
        }, props);
    }

    /*
     *  name can include optional attributes::
     *
     *     play:fw,3x,li,spin,rotate-90
     *
     *  fw:             fixed witdth icon (e.g. for menus)
     *  lg,2x,3x,4x,5x: larger size
     *  border:         added border
     *  li:             list item bullet
     *  spin:           add css3 spinning animation
     *  rotate-90/180/270:
     *  flip-horizontal:
     *  flip-vertical
     */
    make_icon(name, style, item) {
        name = name || "";  // catch undefined
        let i_tag = document.createElement('i');
        let nameparts = name.split(':');
        name = nameparts[0];

        dom.add_classes(i_tag, ...this.classes);
        let classname = this.prefix + (this[name] !== undefined ? this[name] : name);
        i_tag.classList.add(classname);
        if (name) i_tag.classList.add(name);
        if (nameparts.length > 1) {
            let attrs = nameparts[1].split(',');
            attrs.forEach(attr => {
                if (this[attr.replace('-', '_')]) {
                    i_tag = this[attr.replace('-', '_')](i_tag);
                } else {
                    // res.addClass(this.prefix + attr);
                    i_tag.classList.add(this.prefix + attr);
                }
            });
        }

        // if (style) res.css(style);

        return i_tag;
    }
    
    static fontawsome4() {
        return new IconLibrary({
            classes: ['icon', 'fa'],
            prefix: 'fa-',
            url: "https://static.datakortet.no/font/fa470/css/font-awesome.css",
            // save: 'save',
            remove: 'trash-o',
            cancel: 'times-circle-o',
            up: 'chevron-up',
            down: 'chevron-down'
            // plus, minus, play, paperclip
        });
    }
}


export const icon = (function () {
    let chosen_icons = IconLibrary.fontawsome4();
    // dkrequire(chosen_icons.url);
    return (...args) => chosen_icons.make_icon(...args);
}());


// called from create_dk_package.js
export function jq_dkicons(dk) {
    dk.$(document).ready(function () {
        $('[dk-icon]').each(function () {
            $(this).append(icon($(this).attr('dk-icon')));
        });
    });
}


class DK_ICON extends HTMLElement {
    constructor() {
        super();
        this.default = {
            classes: ['icon', 'fa'],
            prefix: 'fa-',
            url: 'https://static.datakortet.no/font/fa470/css/font-awesome.css',
            icons: {
                remove: 'trash-o',
                cancel: 'times-circle-o',
                up: 'chevron-up',
                down: 'chevron-down'
            }
        };
        this._value = null;
        this._name = null;
        this._modifiers = [];
        this._tag = null;
    }

    _icon_stylesheet() {
        return `<link rel="stylesheet" href="${this.default.url}">`;
    }

    _remove_value() {
        if (!this._tag && this._name) return;
        dom.remove_classes(
            this._tag,
            this._name, ...this.default.classes, this.default.prefix + this._name, ...this._modifiers
        );
    }

    _add_value(name, modifiers) {
        if (!name) return;
        if (this.default.icons[name]) name = this.default.icons[name];
        this._name = name;
        this._modifiers = modifiers;
        dom.add_classes(
            this._tag,
            this._name, ...this.default.classes, this.default.prefix + name, ...modifiers
        );
    }

    _spec2modifiers(spec) {
        return spec.split(',').map(attr => this.default.prefix + attr);
    }

    _value2args(value) {
        let [name, ...modifier_list] = (value || "").split(':');
        const modifier_spec = modifier_list.join(',');              // string
        const modifiers = this._spec2modifiers(modifier_spec);
        return [name, modifiers];
    }

    _change_value(value) {
        this._value = value;
        if (!this._tag) return;
        this._remove_value();
        this._add_value(...this._value2args(value));
    }

    connectedCallback() {
        // this.root = this.attachShadow({mode: 'open'});
        // this.root.innerHTML = this._icon_stylesheet();
        // this.root.innerHTML = `<link rel="stylesheet" href="${this.default.url}">`;
        this._tag = document.createElement('i');
        // this.root.appendChild(this._tag);
        this.appendChild(this._tag);
        if (this.value != null) {
            this._add_value(...this._value2args(this.value));
        }
    }

    invalidate() {
        this.icon.setAttribute('value', this.value);
    }

    attributeChangedCallback(attrname, oldval, newval) {
        if (attrname === 'value') {
            this._change_value(newval);
        }
    }

    static get observedAttributes() {
        return ['value'];
    }

    get value() {
        return this._value;
    }

    set value(val) {
        if (val !== this._value) {
            this._change_value(val);
            this.setAttribute('value', val);
        }
    }
}

if (typeof customElements !== 'undefined') customElements.define('dk-icon', DK_ICON);
