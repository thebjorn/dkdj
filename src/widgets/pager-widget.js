
// const $ = require('jquery');
// const dk = require('../boot');
// const Widget = require('../widgetcore/dk-widget.js');

import dk from "../dk-obj";
import {Widget} from "../widgetcore/dk-widget";
import {dkconsole} from "../lifecycle/dkboot/dk-console";

export class PagerWidget extends Widget {
    constructor(...args) {
        super({
            // type: 'PagerWidget',

            pagecount: 0,               // total number of pages
            curpage: undefined,         // currently selected page (zero-based)

            classes: ['pagination'],
            css: {
                marginTop: 0,
                marginBottom: 0
            },
            template: {root: 'ul'},
            
        }, ...args);
    }
    

    /*
     *  Set the pagecount to `n`.
     */
    set_pagecount(n) {
        if (this.pagecount !== n) {
            this.pagecount = n;
            this.draw();
        }
    }

    /*
     *   Select page `n` (zero-based).
     */
    select_page(n) {
        if (n === this.curpage || this.pagecount=== 0) return;     // nothing to do
        if (n < 0 || n >= this.pagecount) {
            dkconsole.debug("PagerWidget.select_page: out of bounds ", n, this);
            return;
        }
        this.curpage = n;
        this.draw();
        this.trigger('select-page', n);
    }

    handlers() {
        const self = this;
        this.widget().on('click', 'li.page', function () {
            self.select_page(dk.$(this).data('page'));
        });
        this.widget().on('click', 'li.prev-nav', function () {
            if (self.curpage > 0) {
                self.select_page(self.curpage - 1);
            }
        });
        this.widget().on('click', 'li.next-nav', function () {
            if (self.curpage + 1 < self.pagecount) {
                self.select_page(self.curpage + 1);
            }
        });
    }
    _prerange(a, b) {
        return (b < a)? null: [a, Math.min(b + 1, a + 2)];
    }
    _postrange(a, b) {
        return (a > b)? null: [Math.max(a, b - 1), b];
    }
    draw_range(range) {
        for (let i = range[0]; i <= range[1]; i++) {
            const li = dk.$('<li/>', { page: i }).addClass('page').data('page', i-1);
            li.append(dk.$('<a href="#"/>').text(i));
            if (i === this.curpage + 1) {
                li.addClass('active');
            }
            this.widget().append(li);
        }
    }
    draw() {
        this.widget().empty();
        this.widget().append(dk.$('<li class="prev-nav"><a href="#">&laquo;</a></li>'));

        if (this.pagecount > 0) {
            let prerange = this._prerange(1, this.curpage - 3);
            let midrange = [Math.max(1, this.curpage - 3), Math.min(this.pagecount, this.curpage + 3)];
            let postrange = this._postrange(this.curpage + 3, this.pagecount);

            // console.log("PRE:", prerange);
            // console.log("MID:", midrange);
            // console.log("POST:", postrange);
            // console.log('----after...---');
            
            if (prerange && prerange[1] + 1 >= midrange[0]) {
                midrange[0] = 1;
                prerange = null;
            }
            if (postrange && midrange[1] + 1 >= postrange[0]) {
                midrange[1] = postrange[1];
                postrange = null;
            }
            // console.log("PRE:", prerange);
            // console.log("MID:", midrange);
            // console.log("POST:", postrange);
            if (prerange) {
                this.draw_range(prerange);
                this.widget().append(dk.$('<li class="ellipsis"><span>...</span></li>'));
            }
            if (midrange) {
                this.draw_range(midrange);
            }
            if (postrange) {
                this.widget().append(dk.$('<li class="ellipsis"><span>...</span></li>'));
                this.draw_range(postrange);
            }
        }

        this.widget().append(dk.$('<li class="next-nav"><a href="#">&raquo;</a></li>'));
    }

}
