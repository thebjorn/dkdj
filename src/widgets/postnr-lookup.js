
/*
 *  Usage::
 *      PostnrLookupWidget.create_on('#my-form', {
 *          postcode_id: '#id_postcode',
 *          city_id: '#id_city'
 *      });
 */
import {Widget} from "../widgetcore/dk-widget";

export class PostnrLookupWidget extends Widget {
    constructor(...args) {
        super({
            // type: 'postnr-lookup-widget',

            // XXX: move into constructor?
            zipcode: null,  // url paramter
            postcode_id: null,
            city_id: null,

            url: '//cache.norsktest.no/ajax/poststed/<%= zipcode %>/',
            urldata: {
                zipcode: function () { return this.zipcode || undefined; }
            },
        }, ...args);
    }

    construct() {
        this.postnr = this.widget(this.postcode_id);
        this.poststed = this.widget(this.city_id);
        this.poststed.attr('tabindex', 7);
        this.zipcode = this.postnr.val();
    }

    draw(poststed) {
        this.poststed.val(poststed || '');
    }

    handlers() {
        const self = this;

        self.postnr.blur(function () {
            self.zipcode = self.postnr.val();
            self.refresh();
        });
    }
}
