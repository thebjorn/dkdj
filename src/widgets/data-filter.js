

/*
 *   This is the top-level filter widget, which groups all sub-filters
 *   for a given data table/grid.
 */
// const _ = require('lodash');
// const dk = require('../boot');
// const Widget = require('../widgetcore/dk-widget.js');
// const forms = require('../forms');

import dk from "../dk-obj";
import {Widget} from "../widgetcore/dk-widget";
import {Template} from "../browser/dk-dom-template";
import {CheckboxSelectWidget, RadioSelectWidget, TriboolWidget} from "../forms/widgets";
import {zip_object} from "../collections";


export function list2data (lst) {
    if (Array.isArray(lst)) {
        const res = {};
        lst.forEach(function (item) {
            res[item] = item;
        });
        res.__order = lst;      // XXX: Not needed anymore
        return res;
    } else {
        return lst;
    }
}


/*
 *  The top level filter container/widget.
 *
 *  dk.filter.DataFilter.create_on(filterbox, {
 *      ...
 *      filters: {
 *          testleader: {
 *              label: 'Testleder',
 *              select_multiple: true,
 *              url: 'progress/all!get-testleader-values'
 *          },
 *          favyear: {
 *              label: 'Favoritt√•r',
 *              values: [2013, 2014, 2015, 2016],
 *              select_multiple: true
 *          },
 *          smile: {
 *              label: 'Smil',
 *              value: undefined,
 *              construct(location) {
 *                  const self = this;
 *                  const btn = $('<button/>').text(":-)");
 *                  this.widget = location.append(btn);
 *                  $(this.widget).on('click', function () {
 *                      self.value = "*SMILE*";
 *                      dk.publish(self.widget, 'change', self.widget);
 *                  });
 *              }
 *          }
 *      }
 *  });
 */
export class DataFilter extends Widget {
    constructor(...args) {
        super({
            // type: 'DataFilter',
            data: undefined,
            heading: 'Filter',
            filters: [],         // XXX shouldn't this be {} or null?
            dataset: null,
            structure: {
                header: {
                    title: {
                        text: 'Filter'
                    }
                },
                content: {}
            },

            template: {
                root: 'section',
                header: '<header/>',
                title: '<h2/>',
                content: '<div/>',

                // Each filter is contained in a .filterbox that is structured
                // similarly to the main filter section.
                filterbox: '<div/>',
                filterheader: '<header/>',
                filtertitle: '<h4/>',
                filtercontent: '<div class="filter content"/>'
            },
            
        }, ...args);
    }

    /*
     *  construct_filterbx is called for each filter definition in this.filters
     *  and creates the radio/check boxes for the filters, calls the data-set
     *  to fetch all filter values, etc.
     */
    construct_filterbx(location, filter, filter_name) {
        const self = this;
        const structure = {
            filterbox: {
                classes: [filter_name],
                filterheader: {
                    filtertitle: {
                        text: filter.label || filter_name
                    }
                },
                filtercontent: {}
            }
        };
        const templ = Template(structure, this.template);
        const dom = templ.append_to(location, this, location);

        if (!filter.construct) {
            // Filters that don't construct themselves are constructed either as single or
            // multiple selects. Default is single-selection (i.e. radio selects).
            let widgetclass = filter.select_multiple ? CheckboxSelectWidget : RadioSelectWidget;
            if (filter.tribool) {
                widgetclass = TriboolWidget;
            }

            filter.construct = function (loc) {
                // this here refers to the filter, which we are attaching a new method to..
                const me = this;

                // noinspection JSPotentiallyInvalidUsageOfClassThis
                this.widget = widgetclass.create_inside(loc, {
                    data: list2data(this.values),
                    name: filter_name,
                    label: me.label
                });

                if (!this.widget.data) {
                    self.dataset.get_filter_data(filter_name, this.widget.FN('draw'));
                }
            };
        }
        filter.construct(location.filterbox.filtercontent);
        if (filter.widget) {
            // filters with their own `construct()` don't necessarily have `.widget`
            dk.on(filter.widget, 'change').run(function () {
                dk.debug('filter-change', self.values());
                if (self.data) self.data.set_filter(self.values());
                dk.publish(self, 'filter-change', self.values());
            });
        }
        return dom;
    }

    construct() {
        this.filters.forEach((value, key) => {
            //console.log('key:', key, 'value:', value);
            this.construct_filterbx(this.content, value, key);
        });
    }

    values() {
        const self = this;
        const keys = Object.keys(this.filters);
        const vals = keys.map(function (filter_name) {
            const filter = self.filters[filter_name];
            if (filter.value) {
                // filter has defined a value on itself..
                return filter.value;
            }
            if (filter.widget && filter.widget.jquery) {
                // found a jQuery object, get its .val()
                return filter.widget.val();
            }
            if (filter.widget && filter.widget.value) {
                // found a dk.form object
                return filter.widget.value;
            }
        });
        // dk.dir(_.object(keys, vals));
        return zip_object(keys, vals);
    }
}
