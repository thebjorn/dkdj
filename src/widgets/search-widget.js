/*
    Default layout::

        <form class="search-form form-inline" role="form">
            <div class="input-group input-sm">
                <input accesskey="s" name="q" class="form-control" placeholder="Søk.." type="search">
                <span class="input-group-btn">
                    <button class="btn btn-default">
                        <i class="fa fa-times"></i>
                    </button>
                    <button class="btn btn-primary">
                        <i class="fa fa-search"></i>
                    </button>
                </span>
            </div>
        </form>

 */

import dk from "../dk-obj";
import {Widget} from "../widgetcore/dk-widget";
import {icon} from "./dk-icon-library";


export class SearchWidget extends Widget {
    constructor(...args) {
        super({
            // type: 'search',
            classes: ['search-form', 'form-inline'],
            accesskey: 's',
            name: 'q',
            tabindex: undefined,
            search_text: 'Søk',
            go_button_text: icon('search:fw'),
            clear_button_text: icon('times:fw'),

            structure: {
                form_group: {
                    input_group: {
                        search_input: {},
                        clear_button: {},
                        go_group: {
                            go_button: {
                                template: '<button class="go-button btn btn-primary"/>'
                            }
                        }
                    }
                }
            },

            template: {
                root: 'form',
                form_group: '<div class="form-group has-feedback"/>',
                input_group: '<div class="input-group input-group-sm"/>',
                search_input: '<input type="search" class="form-control"/>',
                go_group: '<div class="input-group-btn"/>',
                //go_button: '<button class="go-button btn btn-primary"/>',
                clear_button: '<span class="clear-button form-control-feedback"/>'
            },
            
        }, ...args);
        this._lastval = null;
    }

    construct() {
        const input_group = this.form_group.input_group;

        this.clear_button = input_group.clear_button;
        this.clear_button.append(this.clear_button_text);

        this.go_button = input_group.go_group.go_button;
        this.go_button.append(this.go_button_text);

        this.input = input_group.search_input;
        this.input.attr({
            accesskey: this.accesskey,
            name: this.name,
            tabindex: this.tabindex,
            placeholder: this.search_text + '..'
        });

        this.clear_button.hide();
    }

    // XXX: this widget currently stores state data in a DOM property... 
    get_state() {
        return this.get_value();
    }

    set_state(s) {
        this.set_value(s);
    }

    set_value(v) {
        this.input.val(v);
    }

    get_value() {
        return this.input.val();
    }

    handlers() {
        const self = this;

        const search = function (e) {
            e.stopPropagation();
            if (self._searching) return false;
            self._searching = true;
            dk.trigger(self, 'search', self.get_value());
            self._searching = false;
            return false;
        };

        this.widget('.search-form').on('submit', search);

        this.widget().on('click', '.clear-button', function (e) {
            e.stopPropagation();
            self.set_value("");
            self.trigger("clear");
        });

        this.widget().on('click', '.go-button', search);

        this.input.on('keypress', function (e) {
            if (e.which === 13) {
                return search(e);
            }
        });

        this.input.on('keyup', function () {
            const val = self.input.val();
            if (self._lastval === val) return;
            self._lastval = val;
            if (val === "") {
                self.clear_button.hide();
            } else {
                self.clear_button.show();
            }
            self.trigger("change", val);
        });
    }
}
