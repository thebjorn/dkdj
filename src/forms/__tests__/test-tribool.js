
import $ from 'jquery';
import page from "../../widgetcore/dk-page";
import {TriboolWidget} from "../widgets";


test("tribool", () => {
    document.body.innerHTML = `
        <div id="work">
            <label>hello world</label>
        </div>
    `;
    const work =  $('#work');
    page.initialize(document);
    
    const w = TriboolWidget.create_inside(work);
    
    expect(document.getElementById('work')).toMatchSnapshot();
});
