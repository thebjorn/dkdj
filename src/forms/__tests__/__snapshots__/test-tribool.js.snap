// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tribool 1`] = `
<div
  id="work"
>
  
            
  <label>
    hello world
  </label>
  
        
  <div
    class="TriboolWidget"
    id="tribool-widget-1"
  >
    <div
      class="tribool-label_box"
      id="dk-bx-1"
    />
    <div
      class="true-widget_box"
      id="dk-bx-2"
      style="float: left;"
    />
    <div
      class="false-widget_box"
      id="dk-bx-3"
      style="float: left;"
    />
  </div>
</div>
`;
